### ISSUE TITLE
Cursor initial size/position wrong

### ISSUE THREAD

**Issue by bi4k8** [labwc/labwc#681]
My `~/.config/labwc/autostart` executes `kanshi`, which sets output scale to 2. When I start labwc on a TTY, the cursor shows up near the top-left of the screen, but is half the size it should be (i.e., it is rendered at scale 1). When I move the mouse, the cursor jumps to a position twice as far down and right as it was, and renders at the expected (2x) size.

I believe there are two bugs here, as if I change the call to `wlr_cursor_move(server->seat.cursor, NULL, 0, 0);` to pass any nonzero number for either argument, then the cursor no longer jumps (but is still the wrong size until moved).

**Comment by bi4k8** [comment]
Furthermore, if I comment out this early-return from `wlr_output_cursor_move` in wlroots' `src/output/cursor.c`, the cursor also no longer jumps (but is still the wrong size until moved):
```C
	if (cursor->x == x && cursor->y == y) {
		return true;
	}
```

So I think this part of the bug is a case of wlroots being over-eager in this early-return optimization.

**Comment by bi4k8** [comment]
Hmm, it seems like the cursor size part of this bug may also have some (possibly confounding) interaction with swaybg, as when I run labwc with no autostart programs (whereas prior I was running swaybg) and later manually execute kanshi or wlr-randr to change output scale, the cursor never changes size except when over a client, and reverts to scale=1 size whenever over the (blank) desktop.

**Comment by Consolatis** [comment]
hm.. The cursors are loaded with a scale argument via `wlr_xcursor_manager_load()` in two places:
- `src/cursor.c` `cursor_init()` (initialization, hardcoded to `1`)
- `src/output.c` `handle_output_manager_apply()` (output config change, one call per output, using that output's scale)

> the cursor never changes size except when over a client

In wayland, clients always render their own cursor so this is somewhat expected.

> and reverts to scale=1 size whenever over the (blank) desktop

When the cursor is over the desktop (or SSD or some labwc menu) its labwc that sets the cursor, so that sounds like we are somehow missing to update the cursor scale. AFAIR wlroots caches the cursor images for different scales, we might miss something to tell the cursor manager to use the new scale.

`WLR_NO_HARDWARE_CURSORS=1` might also play a role in this.

**Comment by bi4k8** [comment]
I figured out the upstream bug and its [fix](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/3948) has just been merged (should release in wlroots 0.16.1). Now, the cursor size changes upon first movement but it does not suddenly jump in position.

**Comment by Consolatis** [comment]
Very nice catch there regarding the missing scaling for the early return.

**Comment by bi4k8** [comment]
> Hmm, it seems like the cursor size part of this bug may also have some (possibly confounding) interaction with swaybg, as when I run labwc with no autostart programs (whereas prior I was running swaybg) and later manually execute kanshi or wlr-randr to change output scale, the cursor never changes size except when over a client, and reverts to scale=1 size whenever over the (blank) desktop.

After more testing, I've realized that the relevant factor here is not swaybg, but is in fact the presence/absence of the `XCURSOR_THEME` environment variable or a cursor theme in the `~/.icons` directory. That is, it seems that if neither of these is properly set up to provide a cursor theme (as would be the case when, for example, running `HOME=/tmp/somerandomdir labwc` like I was to test with no rc.xml and no autostart programs), then labwc ends up displaying an unscaled cursor over the desktop, even with runtime scale changes via wlr-randr.

**Comment by Consolatis** [comment]
Did some testing nested (which uses an absolute cursor so results may differ to running from tty):

All testing done with `XCURSOR_THEME=Adwaita labwc -s foot` and no kanshi or similar involved so the scale is set to `1`.
Cursor was left over the empty desktop, then the scale was changed by wlr-randr 3 times, from the original startup `1` to `2`, back to `1` and back to `2`.

wlroots patch applied, lastest master:
- `wlr-randr --output WL-1 --scale 2`
  - needs to move over a different labwc surface (e.g. SSD titlebar) to change size
- `wlr-randr --output WL-1 --scale 1`
  - changes size immediately
- `wlr-randr --output WL-1 --scale 2`
  - changes size immediately

wlroots patch applied, latest master + PR 686:
- `wlr-randr --output WL-1 --scale 2`
  - needs to move over a different labwc surface (e.g. SSD titlebar) to change size
- `wlr-randr --output WL-1 --scale 1`
  - changes size immediately
- `wlr-randr --output WL-1 --scale 2`
  - changes size immediately


Now to the jumping in position:
All tests nested with `WLR_NO_HARDWARE_CURSORS=1` set.

- wlroots patch applied:
  - the cursor jumped on every scale change and when moved jumped back

- wlroots patch not applied:
  - the cursor jumped less, I think it only jumped when going from scale `2` to `1` but not when going from `1` to `2` but I am not sure about that.

When running nested with hardware cursor there was no jumping at all, regardless of the wlroots patch.

So to sum up:
- PR #686 didn't change anything for my test case (which doesn't mean it won't change something else).
- At least nested I still see cursor jumping on scale change, although only when run via `WLR_NO_HARDWARE_CURSORS=1`.

**Comment by Consolatis** [comment]
> After more testing, I've realized that the relevant factor here is not swaybg, but is in fact the presence/absence of the `XCURSOR_THEME` environment variable or a cursor theme in the `~/.icons` directory. That is, it seems that if neither of these is properly set up to provide a cursor theme (as would be the case when, for example, running `HOME=/tmp/somerandomdir labwc` like I was to test with no rc.xml and no autostart programs), then labwc ends up displaying an unscaled cursor over the desktop, even with runtime scale changes via wlr-randr.

Good finding. I can totally reproduce this but wonder why it happens. I don't see multiple cursor images in `/usr/share/icons/Adwaita/cursors/` either so they must be scaled somewhere in wlroots, potentially by the included xcursor.

**Comment by bi4k8** [comment]
A note: the cursor *should* jump on scale change, because the entire coordinate system has changed discontinuously. But it should *not* jump discontinuously the first time the mouse is moved *after* changing scale, which is the bug fixed by the wlroots patch.
